### PR TITLE
Image Submission Form

### LINKED ISSUE #6: Save Image Link
#### As a user I want to save a link to an image.

__Story__:

Now that we have an awesome landing page for our users to experience it is time
to introduce our first feature! In general, we want to deliver value to our
customers as fast as possible. As a result, we have decided that we want our
users to be able to save links to images, and immediately see that we have
saved the URL to the image.

One could argue that simply saving the image, and not showing it, provides some
value. Yes, you could split this story here. However, because saving image URLs
and viewing those images are both relatively small additions that work well
together, it is acceptable to complete them together. Note, on the other hand,
that we do not yet want to implement the image index. Its addition adds too
much complexity, so we’ll save that for a future story.

In general, feature addition, or stories, should be split into the smallest
deliverable chunk that provides value to your users with the caveat that it
might make sense to combine two, sometimes more, tiny and related features into
a single story.

__Acceptance criteria__:
- [x] The link is entered through a form.
- [x] After the form is saved, the link is persisted in the database.
- [x] After the form is submitted, the user is redirected to a page displaying
  the image.
- [x] I cannot successfully save an image with an invalid URL.
- [x] An error message is associated with the appropriate input field on
  failure.
- [x] The landing page links to the image link submission form.

__Discussion Topic__:
- [x] What is a valid image URL?

__Dependencies__:
- Deployment Environment

__Note__: Many image sharing sites will permit one to upload images, or
download images when provided a link. For this project we will only work with
the link, i.e., URL, to various images.

__Note__: To understand the database tables created by the `acts_as_taggable` gem, you may find it useful to use a program like the [DB Browser for SQLite](http://sqlitebrowser.org/).

### PR DESCRIPTION
**Pull request by jschofel**
Closes #6 

---

The following is a checklist to remind you of the steps to take when opening a
new PR. The horizontal line above and everything below can be removed from the
pull request description.

- [ ] I have replaced `X` above with the associated issue number.
- [ ] I have changed the issue's label to "pull request".
- [ ] I have assigned the designated reviewer to this PR.
- [ ] I have set the appropriate label on this PR.
- [ ] I have written clear commit messages ([ref](http://chris.beams.io/posts/git-commit/)).

If needed, please review the document on the
[Image Sharing Project](https://sites.google.com/a/appfolio.com/eng/new-hire-resources/engineering-academy/ropes/image-sharing-project)
for information on how labels should be used.

### CODE REVIEW

File: app/controllers/images_controller.rb [hunk 1]
```diff
@@ -0,0 +1,24 @@
+class ImagesController < ApplicationController
+  def show
+    @image = Image.find(params[:id])
```
Comment (grahamotte) [inline app/controllers/images_controller.rb]: You can use strong params here as well `params.require(:id)`

File: app/controllers/images_controller.rb [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+  def new
+    @image = Image.new
+  end
+
+  def create
+    @image = Image.new(image_params)
+    if @image.save
+      redirect_to @image
+    else
+      render 'new'
```
Comment (grahamotte) [inline app/controllers/images_controller.rb]: Best practice to use symbols `render :new` when possible :)

File: app/helpers/images_helper.rb [hunk 3]
```diff
@@ -0,0 +1,2 @@
+module ImagesHelper
```
Comment (grahamotte) [inline app/helpers/images_helper.rb]: is this file needed if it does nothing?

File: app/models/image.rb [hunk 4]
```diff
@@ -0,0 +1,18 @@
+require 'uri'
+
+class URLValidator < ActiveModel::Validator
+  def validate(record)
```
Comment (grahamotte) [inline app/models/image.rb]: You can make this a bit more readable with some guard statements
```
def validate
  return if record.url.start_with?('https://')
  return if record.url.start_with?('http://')
  
  record.errors[:base] << 'Invalid URL'
end
```

File: app/models/image.rb [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+    record.errors[:base] << 'Invalid URL' unless
+        record.url.start_with?('https://') || record.url.start_with?('http://')
+
+    #  URI.parse(record.url)
+    # rescue URI::Error # triggered if invalid URI
```
Comment (grahamotte) [inline app/models/image.rb]: Any reason to leave this here?

File: app/models/image.rb [hunk 6]
```diff
@@ -0,0 +1,18 @@
+require 'uri'
```
Comment (grahamotte) [inline app/models/image.rb]: Is this used?

File: app/views/images/new.html.erb [hunk 7]
```diff
@@ -0,0 +1,25 @@
+<h1>Upload Image</h1>
+<br>
+<%= form_with scope: :image, url: images_path, local: true do |form| %>
+
+  <% if @image.errors.any? %>
```
Comment (grahamotte) [inline app/views/images/new.html.erb]: If you pull `simple_form` into your repo you can get some contextual error handling for free

File: config/routes.rb [hunk 8]
```diff
@@ -1,4 +1,5 @@
 Rails.application.routes.draw do
   # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
   root 'application#home'
+  resources :images
```
Comment (grahamotte) [inline config/routes.rb]: You're defining more resources than you're actually using with this line implicitly, best to only define what you use

File: db/migrate/20200107190051_create_images.rb [hunk 9]
```diff
@@ -0,0 +1,9 @@
+class CreateImages < ActiveRecord::Migration[5.2]
+  def change
+    create_table :images do |t|
+      t.string :url
```
Comment (grahamotte) [inline db/migrate/20200107190051_create_images.rb]: We probably don't ever expect to have images without a url so would be nice to have this be `null: false` at the db level

File: db/schema.rb [hunk 10]
```diff
@@ -10,6 +10,12 @@
 #
 # It's strongly recommended that you check this file into your version control system.
 
-ActiveRecord::Schema.define(version: 0) do
+ActiveRecord::Schema.define(version: 2020_01_07_190051) do
```
Comment (grahamotte) [inline db/schema.rb]: It's totally fine for this project but note for the future: best not to check in the schema since it can lie to your db, generally better to just run the migrations

File: test/controllers/image_form_controller_test.rb [hunk 11]
```diff
@@ -0,0 +1,7 @@
+require 'test_helper'
+
+class ImageFormControllerTest < ActionDispatch::IntegrationTest
+  # test "the truth" do
```
Comment (grahamotte) [inline test/controllers/image_form_controller_test.rb]: Any tests we can add here?

File: test/models/image_test.rb [hunk 12]
```diff
@@ -0,0 +1,7 @@
+require 'test_helper'
+
+class ImageTest < ActiveSupport::TestCase
+  # test "the truth" do
```
Comment (grahamotte) [inline test/models/image_test.rb]: Here too

File: test/fixtures/images.yml [hunk 13]
```diff
@@ -0,0 +1,7 @@
+# Read about fixtures at http://api.rubyonrails.org/classes/ActiveRecord/FixtureSet.html
+
+one:
```
Comment (grahamotte) [inline test/fixtures/images.yml]: At least as as far as appf goes, we usually don't use fixtures as they can get unmaintainable very fast. Check out https://github.com/appfolio/af_exemplar

File: app/controllers/images_controller.rb (hunk 2 above)
Comment (jschofel) [reply app/controllers/images_controller.rb]: fixed in next commit

File: app/controllers/images_controller.rb (hunk 1 above)
Comment (jschofel) [reply app/controllers/images_controller.rb]: fixed in next commit

File: app/helpers/images_helper.rb (hunk 3 above)
Comment (jschofel) [reply app/helpers/images_helper.rb]: I left the helper files because it was generated by rails.  Should the helper files still be scrapped?

File: app/models/image.rb (hunk 4 above)
Comment (jschofel) [reply app/models/image.rb]: added in next commit

File: app/models/image.rb (hunk 5 above)
Comment (jschofel) [reply app/models/image.rb]: removed in next commit

File: app/models/image.rb (hunk 6 above)
Comment (jschofel) [reply app/models/image.rb]: nope.  removed in next commit

File: config/routes.rb (hunk 8 above)
Comment (jschofel) [reply config/routes.rb]: added only restriction

File: db/migrate/20200107190051_create_images.rb (hunk 9 above)
Comment (jschofel) [reply db/migrate/20200107190051_create_images.rb]: added NOT NULL for url

File: app/helpers/images_helper.rb (hunk 3 above)
Comment (grahamotte) [reply app/helpers/images_helper.rb]: Yeah if a file doesn't do anything, then no reason to have it

File: app/views/images/new.html.erb (hunk 7 above)
Comment (jschofel) [reply app/views/images/new.html.erb]: swapped to simple_form

File: test/controllers/image_form_controller_test.rb (hunk 11 above)
Comment (jschofel) [reply test/controllers/image_form_controller_test.rb]: tests added

File: test/models/image_test.rb (hunk 12 above)
Comment (jschofel) [reply test/models/image_test.rb]: tests added

File: test/fixtures/images.yml (hunk 13 above)
Comment (jschofel) [reply test/fixtures/images.yml]: removed fixture file + from tests

File: app/helpers/images_helper.rb (hunk 3 above)
Comment (jschofel) [reply app/helpers/images_helper.rb]: scrapped files
